src/sage/features/threejs.py

```diff
@@ -0,0 +1,63 @@
+import os
+
+from . import StaticFile
+
+
+class Threejs(StaticFile):
+    r"""
+    A :class:`~sage.features.Feature` which describes the presence of
+    threejs-sage in a few standard locations.
+
+    EXAMPLES::
+
+        sage: from sage.features.threejs import Threejs
+        sage: bool(Threejs().is_present())  # needs threejs
+        True
+    """
+
+    def __init__(self):
+        r"""
+        TESTS::
+
+            sage: from sage.features.threejs import Threejs
+            sage: isinstance(Threejs(), Threejs)
+            True
+        """
+        from sage.env import SAGE_SHARE, THREEJS_DIR
+
+        version = self.required_version()
+
+        threejs_search_path = THREEJS_DIR or (
+            os.path.join(SAGE_SHARE, "jupyter", "nbextension", "threejs-sage"),
+            os.path.join(SAGE_SHARE, "sagemath", "threejs-sage"),
+            os.path.join(SAGE_SHARE, "threejs-sage")
+            )
```

Comment:
I love the design idea and I am all for it. Here is more of a comment that we may want to change some the packaging we have for packages like threejs to match better one of those sensible folders out of the box.
I am installing threejs-sage-extension straight from pypi, 0 patch or changes and this is what I get
```
/usr/share/sage
/usr/share/sage/threejs
/usr/share/sage/threejs/r122
/usr/share/sage/threejs/r122/three.min.js
/usr/share/sage/threejs/version
```
We are still at `sage` instead of `sagemath`. I would love for the package to install in the jupyter nbextension folder out of the box which feels right and sensible. Right now, none of these paths, that are looking sensible, will work out of the box without a little bit of work on threejs-sage-extension.

Reply:
Good point, so:
 1. I'll add `SAGE_SHARE/sage/threejs` to the search path above, does that sound right?
 2. I agree that it would make more sense for the threejs sage extension to be installed into jupyter folder, i.e. as above `SAGE_SHARE/jupyter/nbextensions/threejs-sage`.
 3. In the same  I think something like this could eventually be done:
```diff
--- a/build/pkgs/threejs/spkg-install.in
+++ b/build/pkgs/threejs/spkg-install.in
@@ -1,2 +1,2 @@
-sdh_install    src/version "${SAGE_SHARE}/threejs-sage/"
-sdh_install -T src/build   "${SAGE_SHARE}/threejs-sage/$(cat src/version)"
+sdh_install    src/version "${SAGE_SHARE}/jupyter/nbextensions/threejs-sage/"
+sdh_install -T src/build   "${SAGE_SHARE}/jupyter/nbextensions/threejs-sage/$(cat src/version)"
```

I wonder if it would be possible for sagemath to use https://pypi.org/project/pythreejs/ instead of this custom threejs.

Reply:
@kiwifb Where is your `threejs-sage-extension` coming from? The one I found in pypi is named `jupyter-threejs-sage`. Current version is 123 and using pip to install fails with
 
   FileNotFoundError: [Errno 2] No such file or directory: 'version'

There doesn't seem to be version 122, what worked for me is to install `jupyter-threejs-sage=122b1` which installed `.../share/jupyter/nbextensions/threejs-sage/r122/three.min.js`.

Reply:
Your comments surprised me so I doubled checked. I indeed use 122b1. You did install using pip? I will inspect what happens when I install.

Reply:
I made a stupid mistake. For some reason I was looking at the files of `threejs-sage-extension`. Wrong package. Sometimes, I am wondering if I becoming senile. Now the curious bit to me is that I do not have `threejs-sage-extension` installed and everything doctest fine.

Reply:
Indeed:
```
$ python -m venv threejs-sage
$ . threejs-sage/bin/activate
(threejs-sage) $ pip install jupyter-threejs-sage==122b1
Collecting jupyter-threejs-sage==122b1
  Using cached jupyter_threejs_sage-122b1-py3-none-any.whl
Installing collected packages: jupyter-threejs-sage
Successfully installed jupyter-threejs-sage-122b1
(threejs-sage) $ pip show --files jupyter-threejs-sage
Name: jupyter-threejs-sage
Version: 122b1
...
Files:
  ../../../share/jupyter/nbextensions/threejs-sage/r122/three.min.js
  jupyter_threejs_sage-122b1.dist-info/INSTALLER
  jupyter_threejs_sage-122b1.dist-info/LICENSE
  jupyter_threejs_sage-122b1.dist-info/METADATA
  jupyter_threejs_sage-122b1.dist-info/RECORD
  jupyter_threejs_sage-122b1.dist-info/REQUESTED
  jupyter_threejs_sage-122b1.dist-info/WHEEL
  jupyter_threejs_sage-122b1.dist-info/top_level.txt
  jupyter_threejs_sage/__init__.py
  jupyter_threejs_sage/__pycache__/__init__.cpython-312.pyc
  jupyter_threejs_sage/static/r122/three.min.js
(threejs-sage) $ 
```

Note that:
 - pipy only has 122b1 and 123 (the latter is broken)
 - https://github.com/sagemath/threejs-sage/ shows `r130` as the latest release (jul 7, 2021)
 - repology claims `r124` is the newest version for https://repology.org/project/threejs-sage/ and `r122` for https://repology.org/project/threejs/
 - updating is a pain since the filepaths are versioned (wtf?)

:shrug:

Reply:
I am also increasingly not a morning person. Should re-read the whole thing. Yes I have packages for `threejs-sage-extension` which installs where I said it does, and the installation is manual. But I was mistakenly looking at the ebuild of `jupyter-threejs-sage` which is coming from pypi and would install nicely as an extension.

Which makes me wonder, was one meant to replace the other? It looks like they have payload with the same name, just different location. `/usr/share/sage/threejs/r122/three.min.js` for one and `/usr/share/jupyter/nbextensions/threejs-sage/r122/three.min.js` for the other.

Reply:
And those files are identical too.

Reply:
Ok, so your `threejs-sage-extension` package is just like my `threejs-sage` package and sage-the-distro `threejs` package, and arch `threejs-sage` package, which gets a tarball (presumably from github repo) and places the files by hand in a certain location.  It's just that you place it in `/usr/share/sage/threejs`, I doo in `/usr/share/sagemath/threejs-sage` arch does it in `/usr/share/threejs-sage` and sage-the-distro does it in `SAGE_SHARE/threejs-sage`. And the pypi package looks in jupyter nbextensions directory.

Now all of those are caught by the search path above, and I hope @orlitzky will forgive us for looking into 4 different places for the file. And all this trouble is very relevant to our discussion about external packages that are created without taking into account the cost of future maintenance (which noone does, so the maintenance cost carries over to whoever has to sort out the mess when it blows up).

I'll mark this as resolved.

Is this working? Can we flag this as positive review now?

Reply:
It's not about the number of locations. The ThreeJS data is unusable without sage and sage is unusable without the ThreeJS data. Packaging either individually becomes awkward, and when you do package them, they're coupled because they have to agree on the data directory.

Contrast with a python package where the search path is implicitly determined by the --libdir used to configure python. Neither sage nor the database package needs to know that libdir for them to work together. Sage depends on conway-polynomials, but conway-polynomials is now completely independent of sage. The dependency graph has become simpler.

I've been trying to get Sage into Gentoo for 15 years and I'd really like to be able to cross these database packages off my TODO list one-at-a-time. Making them independently useful and of high quality makes that easier, practically and politically.

Reply:
> It's not about the number of locations. The ThreeJS data is unusable without sage and sage is unusable without the ThreeJS data. Packaging either individually becomes awkward, and when you do package them, they're coupled because they have to agree on the data directory.

That's not my point: there actually is a python package for threejs, under the sagemath organization, and that is a mess, so much that sage-the-distro is not using it.

Moreover, in this case the data directory is not the worst problem, since the package has a versioned directory.


> Contrast with a python package where the search path is implicitly determined by the --libdir used to configure python. Neither sage nor the database package needs to know that libdir for them to work together. Sage depends on conway-polynomials, but conway-polynomials is now completely independent of sage. The dependency graph has become simpler.

Sure, that's nice. But then you do cremona elliptic curves database, and place the files in a python package. Now pari/gp also uses this data, and they like perl more than python, so let us also do a perl package so the database can be found, etc. Now I end up with several copies of the database, several external packages, etc.

> I've been trying to get Sage into Gentoo for 15 years and I'd really like to be able to cross these database packages off my TODO list one-at-a-time. Making them independently useful and of high quality makes that easier, practically and politically.

Sure, for *some* of the datasets, having a high quality independent python package sounds very useful. Also a lot of work to get it right. Now, when you say "cross the database packages off your TODO list", what do you exactly mean? For me it was one of the easier parts of packaging sagemath. I decided that I'd place them in `/usr/share/sagemath` and it's just a matter of dumping files in there. Up until this PR I had to setup `SAGE_SHARE`, now they will be autodetected and that's it. It ain't broken, don't fix it. Changing these packages for new packages (python or not) costs a lot of churn for little or no gain. Unless, of course, you come up with a way to share the cremona database between sagemath and pari/gp, and a general way to manage databases, etc. Ideally something that upstream (= source of the data) can manage and maintain on their own, so I actually get a database package from upstream, just like we get ipython from upstream.

Reply:
> Sure, that's nice. But then you do cremona elliptic curves database, and place the files in a python package. Now pari/gp also uses this data, and they like perl more than python, so let us also do a perl package so the database can be found, etc. Now I end up with several copies of the database, several external packages, etc.

In the worse case (no sharing), you get the same number of copies as if you'd bundled everything with sage.

The database packages don't have to be python packages if that's not the best option. In your example, if we ship the raw data with only a pkg-config file and "build" it with autotools, then both Sage and PARI can query `$datadir` using pkg-config. If the interface is more complicated and has multiple consumers, we could write it in C as opposed to python.

> Sure, for _some_ of the datasets, having a high quality independent python package sounds very useful. Also a lot of work to get it right. Now, when you say "cross the database packages off your TODO list", what do you exactly mean? For me it was one of the easier parts of packaging sagemath. I decided that I'd place them in `/usr/share/sagemath` and it's just a matter of dumping files in there. Up until this PR I had to setup `SAGE_SHARE`, now they will be autodetected and that's it. It ain't broken, don't fix it.

As they are now, these packages are useless without Sage, and depend on paths used by Sage. But we don't have Sage in Gentoo, and it could still be a while before we do. Gentoo has hundreds of developers with teams dedicated to various package ecosystems (e.g. math packages), a QA process, two or three permanently-running tinderboxes, tree cleaners, distfile mirrors, binary package builders...

If I were to add these database packages now, it would be adding unusable junk to the tree that would create work for all of those people, potentially for years. Technically I could do it (anyone can commit a new package), but I would rather not spend my political capital that way, especially for something that I don't think is a great idea in the first place.